### PR TITLE
Use safe softmax for masked MHA weights

### LINKED ISSUE #160064: MultiheadAttention returns all NaN when using attn_mask to fully mask some heads, and using `need_weights=True`
### 🐛 Describe the bug

I am trying to use `MultiheadAttention` with `attn_mask`, where I need to fully mask a head, so the boolean mask is False for all the sequence dimension of a given head. The appears to work OK when I set `need_weights=False`, however, when I set `need_weights=True`, I get all NaNs, and the weights are NaN only for the head where the mask is False. Here is an example script and the result I get: 

```python
import torch
from torch import nn

x = torch.randn(1, 5, 3)
mask = torch.tensor([False, False, True]).view(-1, 1, 1).expand(-1, 5, 5)
att = nn.MultiheadAttention(3, 3, batch_first=True)
y1, _ = att(x, x, x, attn_mask=mask, average_attn_weights=False, need_weights=False)
y2, weights = att(x, x, x, attn_mask=mask, average_attn_weights=False, need_weights=True)

print("\nWithout weights:")
print("================")
print(y1)


print("\nWith weights:")
print("==============")
print(y2)
print("weights:")
print("---------")
print(weights)
```

and the results:

```python
Without weights:
================
tensor([[[ 0.0446, -0.0386,  0.0045],
         [ 0.0430, -0.0491, -0.0094],
         [ 0.0600, -0.0517,  0.0065],
         [ 0.0281, -0.0649, -0.0441],
         [ 0.0531, -0.0635, -0.0148]]], grad_fn=<TransposeBackward0>)

With weights:
==============
tensor([[[nan, nan, nan],
         [nan, nan, nan],
         [nan, nan, nan],
         [nan, nan, nan],
         [nan, nan, nan]]], grad_fn=<TransposeBackward0>)
weights:
---------
tensor([[[[0.1603, 0.2289, 0.1342, 0.3084, 0.1681],
          [0.1755, 0.2203, 0.1567, 0.2665, 0.1809],
          [0.2040, 0.1958, 0.2082, 0.1892, 0.2029],
          [0.1831, 0.2150, 0.1690, 0.2459, 0.1870],
          [0.2166, 0.1802, 0.2374, 0.1545, 0.2113]],

         [[0.2173, 0.1916, 0.2304, 0.1576, 0.2031],
          [0.2078, 0.1967, 0.2131, 0.1807, 0.2017],
          [0.2153, 0.1928, 0.2266, 0.1624, 0.2029],
          [0.1857, 0.2043, 0.1777, 0.2368, 0.1955],
          [0.2015, 0.1994, 0.2025, 0.1962, 0.2004]],

         [[   nan,    nan,    nan,    nan,    nan],
          [   nan,    nan,    nan,    nan,    nan],
          [   nan,    nan,    nan,    nan,    nan],
          [   nan,    nan,    nan,    nan,    nan],
          [   nan,    nan,    nan,    nan,    nan]]]], grad_fn=<ViewBackward0>)
```

After doing some investigation, I found that the source of the error comes from this [line](
https://github.com/pytorch/pytorch/blob/38d65c64658928929a5c70114b56041096aaf0dd/torch/nn/functional.py#L6452), since all the entries are -infinity, softmax returns NaN. 

When `need_weights=False`, the code uses the native implementation of `scaled_dot_product_attention`, which is calculating the softmax safely. I haven't tested this working part, but I think it is running this [code](https://github.com/pytorch/pytorch/blob/38d65c64658928929a5c70114b56041096aaf0dd/aten/src/ATen/native/transformers/attention.cpp#L911) which is checking if a complete row is masked, return all zeros instead of the softmax result. This same logic should be replicated in the python code which is run when `need_weights=True`...

### Versions

```bash
Collecting environment information...
PyTorch version: 2.7.1+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 15.1.1 20250729
Clang version: 20.1.8
CMake version: version 4.0.3-dirty
Libc version: glibc-2.42

Python version: 3.13.5 (main, Jun 21 2025, 09:35:00) [GCC 15.1.1 20250425] (64-bit runtime)
Python platform: Linux-6.15.8-arch1-1-x86_64-with-glibc2.42
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1080 Ti
Nvidia driver version: 575.64.05
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           39 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  8
On-line CPU(s) list:                     0-7
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz
CPU family:                              6
Model:                                   158
Thread(s) per core:                      2
Core(s) per socket:                      4
Socket(s):                               1
Stepping:                                9
CPU(s) scaling MHz:                      88%
CPU max MHz:                             4200.0000
CPU min MHz:                             800.0000
BogoMIPS:                                7200.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault pti tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp vnmi
Virtualization:                          VT-x
L1d cache:                               128 KiB (4 instances)
L1i cache:                               128 KiB (4 instances)
L2 cache:                                1 MiB (4 instances)
L3 cache:                                8 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-7
Vulnerability Gather data sampling:      Vulnerable: No microcode
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             KVM: Mitigation: Split huge pages
Vulnerability L1tf:                      Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:                       Vulnerable: Clear CPU buffers attempted, no microcode; SMT vulnerable
Vulnerability Meltdown:                  Mitigation; PTI
Vulnerability Mmio stale data:           Vulnerable: Clear CPU buffers attempted, no microcode; SMT vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Vulnerable
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Vulnerable
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Retpolines; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Vulnerable: No microcode
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Vulnerable: Clear CPU buffers attempted, no microcode; SMT vulnerable

Versions of relevant libraries:
[pip3] mypy==1.17.0
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.3.1
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] pytorch-lightning==2.5.2
[pip3] torch==2.7.1
[pip3] torchaudio==2.7.1
[pip3] torchmetrics==1.7.4
[pip3] torchvision==0.22.1
[pip3] triton==3.3.1
[conda] Could not collect
```

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185338

When MultiheadAttention runs with need_weights=True, it uses the Python
bmm/baddbmm attention path instead of scaled_dot_product_attention. Boolean
attention masks and key padding masks are canonicalized into additive -inf
scores before this branch. If a mask fully hides a query row or a whole head,
the old plain softmax saw an all--inf row and produced NaNs, which then
poisoned both the returned weights and the projected output.

The need_weights=False path already goes through SDPA, whose math backend uses
safe softmax and returns zero weights for all--inf rows. Mirror that behavior
in the need_weights=True path when a mask is present by using torch._safe_softmax
for masked scores. Leave unmasked scores on the existing softmax path to avoid
changing behavior or cost for the common unmasked case.

I used the existing safe softmax primitive rather than manually checking
isneginf rows in Python, since this keeps the semantics aligned with SDPA. An
unconditional replacement was considered, but the microbenchmarks below showed
that it would slow unmasked need_weights=True calls unnecessarily.

Fixes #160064
Generated by my agent

Benchmark Results:
Direct CPU softmax microbenchmark on masked score tensors:
- shape=(8, 128, 128): plain softmax 85.78 us, _safe_softmax 293.46 us
- shape=(32, 256, 256): plain softmax 1.26 ms, _safe_softmax 4.76 ms

End-to-end CPU MultiheadAttention need_weights=True masked-path benchmark:
- issue shape, batch=1 seq=5 embed=3 heads=3: old 88.61 us, new 98.22 us
- larger case, batch=8 seq=128 embed=128 heads=8: old 4.38 ms, new 7.35 ms

Test Plan:
python - <<'PY'
import torch
from torch import nn

torch.manual_seed(0)
x = torch.randn(1, 5, 3)
mask = torch.tensor([False, False, True]).view(-1, 1, 1).expand(-1, 5, 5)
att = nn.MultiheadAttention(3, 3, batch_first=True)
y1, _ = att(x, x, x, attn_mask=mask, average_attn_weights=False, need_weights=False)
y2, weights = att(x, x, x, attn_mask=mask, average_attn_weights=False, need_weights=True)
print(torch.isnan(y1).any().item(), torch.isnan(y2).any().item(), torch.isnan(weights).any().item())
print(torch.allclose(y1, y2))
print(torch.equal(weights[0, 2], torch.zeros_like(weights[0, 2])))
PY
python test/nn/test_multihead_attention.py TestMultiheadAttentionNN.test_multihead_attn_fully_masked_head_need_weights TestMultiheadAttentionNN.test_multihead_attn_fully_masked_row_need_weights
python test/nn/test_multihead_attention.py -k test_multihead_attention
lintrunner -a

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by drisspg — CHANGES_REQUESTED** [review]
mha is frozen

**Comment by jansel** [comment]
My agent says the maintainer review says nn.mha is frozen and no changes are planned, and the linked issue was closed for that reason. Closing this PR instead of continuing to update it.
